README
	
	vnStat SVG frontend

	1. Introduction
	2. Demo
	3. Usage
	4. Desgin
	5. More...

------------

1. Introduction

    vnStat SVG frontend(vnStatsvg)[1] is another web-based graphical frontend
    to vnStat(a network traffic monitor for Linux)[2]. it is suited to the
    distributed & embedded system built on linux.

2. Demo

   PLEASE use mozilla-5.0 based browsers to browser vnStatSVG:

    http://tinylab.org/vnstatsvg/ is a demo site of it.

    See screenshot pictures under DEMO/ if can not access it.

   * Browsers list supported by the latest vnStatSVG(1.0.0):

	$ seamonkey --version
		SeaMonkey 1.1.9, Copyright (c) 2003-2008 mozilla.org, build 2008031300
	$ firefox --version
		Mozilla Firefox 3.0, Copyright (c) 1998 - 2008 mozilla.org
	$ chromium-browser --version
		Chromium 25.0.1364.160 Ubuntu 12.10

3. Usage

	please read the INSTALL document for standard Linux distribution.
	and INSTALL_Busybox document for Busybox based embedded system.
	To learn multi-protocols configuration, please refer to INSTALL_MultiProtocol

4. Desgin

    it is based on HTML, XML, XSL, Javascript, CSS, SVG, for more information,
    please read the PAPER[5](this is a paper I have published in "The First
    IEEE International Conference on Ubi-media Computing").

    if you have interest, you can try to use another web-based vnStat
    frontendp[3] and compare it with vnStatSVG, and then try to understand why
    vnStatSVG is more suited to distributed & embedded system.

5. More...

   this project is based on the contest[4] i have attended in 2007.

   there is new tool named vnstatgui can be used to generate the graphical
report. it is in src/vnstatgui/.

References & Links

[1] vnStatsvg
http://vnstatsvg.sf.net
[2] vnStat, a network traffic monitor for linux
http://humdi.net/vnstat/
[3] vnStat PHP frontend, another web-based frontend, 
http://www.sqweek.com/sqweek/index.php?p=1
[4] elf: scalable firewall
http://elf.oss.lzu.edu.cn
[5] A CGI+AJAX+SVG based monitoring method for the distributed and embedded
system



ChangeLog

	ChangeLog

	PLEASE add the latest change log on the top of THIS file, and
	follow this guidelines:
		[*] list the change log under the version number
		[*] list the change Time and your Email(P)
		[*] list the description of your change(D)

--------

[*] Version 1.0.0

3-Jul-2008

P: wu zhangjin, <wuzhangjin@gmail.com>
D: Arrange version 0.1.x and release a new version
D: Support multi-hosts via a light-weight http proxy
D: Reduce the SIZE of the XML output via using short XML flags
D: Substitude strftime() in gawk for the date command in shell version
D: Rearrange the XSL file, and write comments to every lines
D: Write vnStatXML to dump the data of vnStat in XML format
D: Use setTimeout/clearTimeout to monitor the Memory information
D: Add the configuration/Installation scripts

22-Oct-2008

P: wu zhangjin, <wuzhangjin@gmail.com>
D: add a tool vnstatgui to generate the graphical output in localhost.
D: it is src/vnstatgui

03-May-2009

P: wu zhangjin, <wuzhangjin@gmail.com>
D: bugfix: fix the parsing of addr:port 

28-Jul-2013

P: wu zhangjin, <wuzhangjin@gmail.com> of TinyLab, tinylab.org
D: bugfix: add busybox awk support
D: Add chromium browser support via fixing up vnstat.xsl and vnstat.js
D: Check httpd, httpd2, awk if no apache2 and gawk exist
D: Fix up compilier optimization of vnstatxml: -O2 -> -Os
D: Fix up linking issue, should use -lm after the other objectives
D: Add cross compiling support
D: Document the usage of local graphic output with vnstatgui
D: Verify and support Busybox based embedded system

28-Jul-2013

P: wu zhangjin, <wuzhangjin@gmail.com> of TinyLab, tinylab.org
D: Add multi protocols support, allows to collect data from remote hosts will any protocols
D: Allows to test the src/cgi-bin/vnstat-{c,p,shell}.sh scripts locally
D: Fix up the output format of the first column


src/admin/vnstat.js
// the Javascript file for vnStatsvg

var vnstat_cgibin="/cgi-bin/";
var vnstat_admin="./";

var vnstat_proxy=vnstat_cgibin+"proxy.sh";
var sidebar_xsl=vnstat_admin+"sidebar.xsl";
var sidebar_xml=vnstat_admin+"sidebar.xml";
var vnstat_xml="";	/* here we will use the CGI program to generate the dynamical XML file */
var vnstat_xsl=vnstat_admin+"vnstat.xsl";

var xmlHttp;
var xmlFile;
var xslFile;
var divID;
var speed = 1000;
var alertTimerId = 0;

function showSidebar()
{
	divID="sidebar"; xmlFile=sidebar_xml; xslFile=sidebar_xsl;
	showXML();
}

function showPage(page, caption)
{
	states = getHTML('status').split(':');
	host = states[0];
	iface = states[1];
	id = states[2];

	ID="SPAN_"+host+id;
	description = document.getElementById(ID).getAttribute("description");
	protocol = document.getElementById(ID).getAttribute("protocol");
	dump_tool = document.getElementById(ID).getAttribute("dump_tool");

	if (protocol == "")
		protocol = "http";
	if (dump_tool == "")
		dump_tool = "/cgi-bin/vnstat.sh";
	if (description == "")
		description = host;

        showHTML("status", host+":"+iface+":"+id);
        showHTML("caption", caption);

        vnstat_xml = dump_tool+"?i="+iface+"&p="+page;
	/* document.domain is the domain name of the "server node", you'd better set 'host' as it */
	/* Note: document.domain is 'undefined' while using firefox to browse busybox httpd, ignore multi-hosts support for such case */
        if ((document.domain && host != document.domain) || protocol != "http") {
                vnstat_xml=vnstat_proxy+"?"+protocol+"://"+host+vnstat_xml;
        }

	// if the bandwidth is "narrow", you can try to uncomment the following line
	showHTML("main_wrapper", "<font color='blue'>Loading...</font>");

	divID="main_wrapper"; xmlFile=vnstat_xml; xslFile=vnstat_xsl;
        showXML();
}

function showMenu(host, iface, id)
{
	if (iface == "")
		iface = "eth0";

        showHTML("status", host+":"+iface+":"+id);

	showHTML("caption", "Click to monitor network traffic.");

	showHTML("main_wrapper", "<p align='center'>Traffic data will be loaded here...</p>");

	hideotherSubmenu(host, id);

	showSubmenu(host, id);
}

function hideotherSubmenu(host, id)
{
	me = host + id;
	spans = document.getElementsByTagName('span');
	for (i=0; i < spans.length; i ++) {
		spanID = spans[i].id;
		spanIDs = spanID.split('_');	/* SPAN_hostid */
		prefix = spanIDs[0];
		divID = spanIDs[1];
		if (prefix == 'SPAN' && divID != me && getHTML(spanID) == '-') {
			clearTimeout ( alertTimerId );
			showHTML(spanID, '+');
			showHTML(divID, ''); 
		}
	}
}

function showSubmenu(host, id)
{
	divID = host+id;
	spanID="SPAN_"+divID;

	if(getHTML(spanID) == '-') {
		clearTimeout ( alertTimerId );
		showHTML(spanID, '+');
		showHTML(divID, ''); 
	} else {
		showHTML(spanID,'-');
		xmlFile="menu.xml"; xslFile="menu.xsl";
		showXML();
	}
}

function getHTML(ID)
{
	return document.getElementById(ID).innerHTML;
}

function showHTML(ID, content)
{
	document.getElementById(ID).innerHTML = content;
	return;
}

function showXML()
{
	clearTimeout ( alertTimerId );

	xmlHttp=GetXmlHttpObject();
	if (xmlHttp == null) return;

	xmlHttp.onreadystatechange = stateChanged;
	xmlHttp.open("GET", xmlFile, true);
	xmlHttp.send(null);
	if(xmlFile.lastIndexOf("second") > 0) {
		alertTimerId=setTimeout("showXML()", speed);
	}
}

function stateChanged()
{
	if (xmlHttp.readyState == 4)
	{
		if (xmlHttp.status == 200 ) {
	    	// Only support Firefox, Opera 8.0+, Safari
	    	var xmlDoc=xmlHttp.responseXML;
		var content;

		if (document.implementation && document.implementation.createDocument) {
			// Chromium, Firefox, Mozilla, Opera, etc.
			//load xsl file
			//var xslDoc =  document.implementation.createDocument("", "", null);
			//xslDoc.async = false;
			//xslDoc.load(xslFile);
			var xslHttp = new XMLHttpRequest();
			xslHttp.open("GET", xslFile, false);
			xslHttp.send();
			var xslDoc = xslHttp.responseXML.documentElement;

			//build the relationship between xml file and xsl file
			var xsl = new XSLTProcessor();
			xsl.importStylesheet(xslDoc);

			var result = xsl.transformToDocument(xmlDoc);
			var xmls = new XMLSerializer();

			// serialize to string
			content = xmls.serializeToString(result);

			// filter the xml header for mozilla/5.0-based firefox 3.0
			if (content.indexOf("<?xml") > -1) {
				var index = content.indexOf(">") + 2;
				var subcontent = content.substr(index);
				content = subcontent;
			}
		} else if(typeof window.ActiveXObject != 'undefined') {
		        //load xsl file
        		xslDoc = new ActiveXObject('Microsoft.XMLDOM');
		        xslDoc.async = false;
	        	xslDoc.load(xslFile);

			content = xmlDoc.documentElement.transformNode(xslDoc);
	    	}
		// append the result
		showHTML(divID, content);
	    } else showHTML(divID, 'Error with XMLHttpRequest!');
	}
}

function GetXmlHttpObject()
{
	var xmlHttp=null;
	try
	{
	  	// Firefox, Opera 8.0+, Safari
		xmlHttp=new XMLHttpRequest();
	}
	catch (e)
	{
		// Internet Explorer: vnStatSVG not support IE currently, Hope one day, IE no long imperious and support AJAX directly!!
		try
		{
			xmlHttp=new ActiveXObject("Msxml2.XMLHTTP");
		}
		catch (e)
		{
			xmlHttp=new ActiveXObject("Microsoft.XMLHTTP");
		}
	}
	return xmlHttp;
}
